mediapipe/graphs/instantmotiontracking/calculators/matrices_manager_calculator.cc

```diff
@@ -0,0 +1,324 @@
+//
+// Example config:
+// node{
+//  calculator: "MatricesManagerCalculator"
+//  input_stream: "ANCHORS:tracked_scaled_anchor_data"
+//  input_stream: "IMU_DATA:imu_data"
+//  input_stream: "USER_ROTATIONS:user_rotation_data"
+//  input_stream: "USER_SCALINGS:user_scaling_data"
+//  output_stream: "MODEL_MATRICES:model_matrices"
+// }
+
+class MatricesManagerCalculator : public CalculatorBase {
+  public:
+    static ::mediapipe::Status GetContract(CalculatorContract* cc);
+    ::mediapipe::Status Open(CalculatorContext* cc) override;
+    ::mediapipe::Status Process(CalculatorContext* cc) override;
+  private:
+    const DiagonalMatrix3f GenerateScalingMatrix(const float scale_factor);
+    const Matrix3f GenerateUserRotationMatrix(const float rotation_radians);
+    const Matrix4fCM GenerateEigenModelMatrix(const Vector3f translation_vector,
+      const Matrix3f rotation_submatrix);
+    const Vector3f GenerateAnchorVector(const Anchor tracked_anchor);
+    const Matrix3f GenerateIMURotationSubmatrix(const float yaw, const float pitch, const float roll);
+
+    // Returns a user scaling increment associated with the sticker_id
+    // TODO: Adjust lookup function if total number of stickers is uncapped to improve performance
+    const float GetUserScaler(const std::vector<UserScaling> scalings, const int sticker_id) {
+      for (const UserScaling &user_scaling : scalings) {
+        if (user_scaling.sticker_id == sticker_id) {
+          return user_scaling.scale_factor;
+        }
+      }
+    }
+    // Returns a user rotation in radians associated with the sticker_id
+    const float GetUserRotation(const std::vector<UserRotation> rotations, const int sticker_id) {
+      for (const UserRotation &rotation : rotations) {
+        if (rotation.sticker_id == sticker_id) {
+          return rotation.rotation_radians;
+        }
+      }
+    }
+
+    // This returns a scale factor by which to alter the projection matrix for
+    // the specified render id in order to ensure all objects render at a similar
+    // size in the view screen upon initial placement
+    const float GetDefaultRenderScale(const int render_id) {
+      if(render_id == 0) { // Robot
+        return 5.0;
+      }
+      else if(render_id == 1) { // Dino
+        return 0.75;
+      }
+      else if(render_id == 2) { // GIF
+        return 0.15;
+      }
+    }
+};
+
+REGISTER_CALCULATOR(MatricesManagerCalculator);
+
+::mediapipe::Status MatricesManagerCalculator::GetContract(
+    CalculatorContract* cc) {
+  RET_CHECK(cc->Inputs().HasTag(kAnchorsTag)
+    && cc->Inputs().HasTag(kIMUDataTag)
+    && cc->Inputs().HasTag(kUserRotationsTag)
+    && cc->Inputs().HasTag(kUserScalingsTag)
+    && cc->InputSidePackets().HasTag(kFOVSidePacketTag)
+    && cc->InputSidePackets().HasTag(kAspectRatioSidePacketTag));
+  RET_CHECK(cc->Outputs().HasTag(kModelMatricesTag));
+
+  if (cc->Inputs().HasTag(kAnchorsTag)) {
```

Comment:
nit: No need for all these HasTag checks anymore, since they're already RET_CHECK'ed above.